src/process/io/StreamSource.cpp

```diff
@@ -0,0 +1,268 @@
+ *                     still stored in the internal rewind buffer. Can e.g. be used by skip.
+ * @param numberOfBytes
+ * @return
+ */
+int64_t StreamSource::read(Bytef *targetBuffer, int numberOfBytes) {
+    if (rewoundBytes > 0) {
+        return readFromBuffer(targetBuffer, numberOfBytes);
+    } else {
+        return readFromStream(targetBuffer, numberOfBytes);
+    }
+}
+
+int64_t StreamSource::readFromBuffer(Bytef *targetBuffer, int numberOfBytes) {
+    char *buf = joinRewindBuffer();
+    int64_t bufLen = getRewindBufferSize();
+
+    int64_t copiedBytes = numberOfBytes;
+    if (numberOfBytes > bufLen) {
+        // Cut it, copy less!
+        copiedBytes = bufLen;
+    }
+
+    if (numberOfBytes > rewoundBytes) {
+        copiedBytes = rewoundBytes;
+    }
+
+    // Copy copiedBytes of data from the right side!
+    if (targetBuffer != nullptr)
+        memcpy(targetBuffer, buf + bufLen - rewoundBytes, copiedBytes); // TODO
+    rewoundBytes -= copiedBytes;
+    currentPosition += copiedBytes;
+
+    delete[] buf;
+
+    return copiedBytes;
+}
+
+int64_t StreamSource::readFromStream(Bytef *targetBuffer, int numberOfBytes) {
+    char *readBuffer = new char[numberOfBytes]{0};
+    inputStream->sync();
+
+    inputStream->read(readBuffer, numberOfBytes);
+    streamsize amountRead = inputStream->gcount();
+
+    if (amountRead > 0) {
+        if (targetBuffer != nullptr)
+            memcpy(targetBuffer, readBuffer, amountRead);// TODO
+        rewindBuffer.emplace_back(readBuffer);
+        rewindBufferEntryLength.emplace_back(amountRead);
+        if (rewindBuffer.size() > maxSegmentsInBuffer) {
+            char *front = rewindBuffer.front();
+            rewindBuffer.pop_front();
+            rewindBufferEntryLength.pop_front();
+            delete[] front;
+        }
+        currentPosition += amountRead;          // Advance current position marker
+    } else {
+        delete[] readBuffer;
+    }
+    return amountRead;
+}
+
+int StreamSource::readChar() {
+    Byte result = 0;
+    int res = static_cast<int>(this->read(&result, 1));
+    return res < 0 ? res : (int) result;
+}
+
+/**
+ * A lot of cases apply here:
+ * - if absolute:
+ *   - nByte > curPos
+ *     => Skip to curPos or eof
+ *   - nByte < curPos
+ *     - nByte < sizeOfBuffer   => Go on
+ *     - nByte > sizeOfBuffer   => Error!
+ *
+ * - else if relative
+ *   - nByte > 0
+ *     => Skip by nByte
+ *   - nByte < 0
+ *     - nByte < sizeOfBuffer
+ *     - nByte > sizeOfBuffer
+ *
+ * To make things easier, convert a relative offset to an absolute offset first. So only the cases for absolute apply
+ *
+ * @param nByte The absolute position in the stream or the relative position to the current position
+ * @param absolute Absolute or relative position
+ * @return -1 for a failure or the new position in the stream
+ */
+int64_t StreamSource::seek(int64_t nByte, bool absolute) {
```

Comment:
Relative and absolute position arithmetic are use case that I tended to solve by dedicated classes (e.g. in Perl code for genomic positions; BTW like 1-based and 0-based position arithmetic). Maybe that could be an option in this system as well.